### ISSUE TITLE
Handling DNSKEY RRs with broken public key field

### ISSUE THREAD

**Issue by mattias-p** [zonemaster/zonemaster#1040]
We have some test cases that specify that the key size of DNSKEY RRs should be calculated. When we receive valid DNSKEY RRs this is not a problem, but when we receive DNSKEY RRs with an empty (or otherwise too short) public key field the situation is unspecified. The key size for these broken DNSKEY RRs simply cannot be calculated.

I could think of three possible solutions to this:
1. Specify that Zonemaster should determine the key size of a DNSKEY RR with insufficient data to be 0.
2. Specify that Zonemaster should ignore DNSKEY RRs with insufficient data.
3. Specify that Zonemaster should ignore DNSKEY RRs with insufficient data, but only when calculating the key size.

Test cases DNSSEC03 and DNSSEC14 explicitly depends on the calculation of the key size. The current implementation of DNSSEC05 also has this problem because it outputs a KEY_DETAILS message with a key size field, apparently for diagnostic reasons. When the public key field is empty the size of the key cannot be calculated. Also most of the DNSSEC test cases send DNSKEY queries so the evaluation of those might change as well depending on the choice of solution. We should probably prioritize maximizing correctness over minimizing change in behavior.

**Comment by matsduf** [comment]
DNSKEY with empty keys are probably very rare, and I suggest that we should handle them in a cheap way, i.e. not spend too much code on them. Ignoring such records is reasonable.

**Comment by mattias-p** [comment]
I'd say we want to avoid Zonemaster giving unreasonable results even in rare cases.

Comparing the implementation time requirement for the above choices, I'd say the difference is most likely negligible.

But if it's your assessment that ignoring DNSKEY records with insufficient data (to calculate the key size) is the solution that gives the most reasonable results across the test cases, I'll go ahead and implement that. I haven't thought deeply about it but this is the option I intuitively like best too.

**Comment by pnax** [comment]
> The key size for these broken DNSKEY RRs simply cannot be calculated.

Instead of discarding such record, can't this be an error message?

**Comment by mattias-p** [comment]
> > The key size for these broken DNSKEY RRs simply cannot be calculated.
> 
> Instead of discarding such record, can't this be an error message?

Like a critical error? I didn't think of that but it makes sense. More thoughts on this?

**Comment by matsduf** [comment]
That would mean that we should add that to every test case that is dependent on DNSKEYs. It is probably extremely rare that DNSKEY is empty. There cannot be any RRSIGs using such DNSKEYs since they cannot be created and cannot be validated. We have since this once.

I strong suggestion is that we just ignore it, and if we in the future see that such a strategy is problematic then we can change it. Else we risk spending energy on something that has extremely little meaning.

**Comment by pnax** [comment]
> Like a critical error? I didn't think of that but it makes sense. More thoughts on this?

Yes, if Zonemaster can catch such inconsistency, it would be nice to explain what happens to the end user (especially since it might be because of a broken record).

> It is probably extremely rare that DNSKEY is empty. There cannot be any RRSIGs using such DNSKEYs since they cannot be created and cannot be validated.

I do think this is the purpose of Zonemaster to be able to catch such rare errors. I agree that we can't cover all edge cases but it it can be nice to add them incrementally (if the roadmap and available resources allow it). 

> That would mean that we should add that to every test case that is dependent on DNSKEYs.

Specification-wise maybe we can find a way to avoid repetition across test cases (even if we don't implement this, this issue of repetition across tests might come back at a later point in time).

**Comment by matsduf** [comment]
> > It is probably extremely rare that DNSKEY is empty. There cannot be any RRSIGs using such DNSKEYs since they cannot be created and cannot be validated.
> 
> I do think this is the purpose of Zonemaster to be able to catch such rare errors. I agree that we can't cover all edge cases but it it can be nice to add them incrementally (if the roadmap and available resources allow it).

I then suggest that we create a specific test case that catches issues like this. We could start with a test case "Check DNS record validity" where selected records are tested, DNSKEY at the start, bu we can add more later.

I all other test cases we just ignore the error.

**Comment by mattias-p** [comment]
> I then suggest that we create a specific test case that catches issues like this. We could start with a test case "Check DNS record validity" where selected records are tested, DNSKEY at the start, bu we can add more later.
>
> I all other test cases we just ignore the error.

I think this is a good approach. While the scope of test cases we have today is a zone, the scope of this kind of test case is a packet (or perhaps a response). (And maybe "scope" isn't the right word to use here? Or maybe it is exactly the right word?) We could group test cases of this kind in a "Packet" module. WDYT?

IIRC we're already making at least one other check for all packages we receive and emitting some message if the check fails. It would be nice to have a place to specify that check as well.

> > Like a critical error? I didn't think of that but it makes sense. More thoughts on this?
> 
> Yes, if Zonemaster can catch such inconsistency, it would be nice to explain what happens to the end user (especially since it might be because of a broken record).

My initial thought was to improve this by enabling Zonemaster to record broken responses like this into its cache. That would be nice because a unit test for the proper behavior when encountering a broken DNSKEY record could be maintained using our standard approach.

The idea of making badly broken responses into an error instead of letting them flow through the test cases is also nice because that way we don't need to keep such responses in mind all the time when specifying and implementing test cases. It would be really neat if any test case that requests such a broken response would be aborted, but independent test cases would be allowed to continue their execution. This way the test run could terminate normally and the cache file could be saved.

By changing the definition of the "critical" severity level and making use of dependencies between test cases, "critical" messages could be such an exception mechanism. We'd just have the test case emit a "critical" message to indicate that any dependent test cases should be skipped (perhaps with a message about why they were skipped). The aborted and skipped test cases would be highlighted with critical messages in the results.

> > It is probably extremely rare that DNSKEY is empty. There cannot be any RRSIGs using such DNSKEYs since they cannot be created and cannot be validated.
> 
> I do think this is the purpose of Zonemaster to be able to catch such rare errors. I agree that we can't cover all edge cases but it it can be nice to add them incrementally (if the roadmap and available resources allow it).

I agree with this. I also think it could be worth putting in slightly more effort early on to establish a sound architecture such that we'll be able to easily add new cases incrementally as we encounter them.

I suggest we do this:
1. Agree on the general outline of such an architecture before resolving this issue. (I laid out a rough idea above in this comment, based on suggestions put forth by you guys.)
2. Resolve this issue by merging an update to the specifications for how to deal with these broken DNSKEY records, that doesn't really do anything to implement the architectural changes, but that is roughly in line with the architecture we have in mind.
3. Merge a refactoring of the specification for the broken DNSKEY behavior into a proper test case, along with any needed changes to the test case specification template, and along with any other changes needed to establish the architecture. Perhaps some new document; who knows, we'll see.
4. As part of our normal processes, incrementally bring existing test case specifications into line with the updated test case specification template.

**Comment by matsduf** [comment]
> I think this is a good approach. While the scope of test cases we have today is a zone, the scope of this kind of test case is a packet (or perhaps a response). (And maybe "scope" isn't the right word to use here? Or maybe it is exactly the right word?) We could group test cases of this kind in a "Packet" module. WDYT?

I do not see the difference in scope. Today there are test cases validating the SOA record in a zone. In this case it would be a test case that would validate the DNSKEY records (and maybe other DNSECE related records) in a zone.

> The idea of making badly broken responses into an error instead of letting them flow through the test cases is also nice because that way we don't need to keep such responses in mind all the time when specifying and implementing test cases. It would be really neat if any test case that requests such a broken response would be aborted, but independent test cases would be allowed to continue their execution. This way the test run could terminate normally and the cache file could be saved.

I think it is better to just ignore the DNSKEY and let the test cases test as much as they can on the data that is available.


> By changing the definition of the "critical" severity level and making use of dependencies between test cases, "critical" messages could be such an exception mechanism. We'd just have the test case emit a "critical" message to indicate that any dependent test cases should be skipped (perhaps with a message about why they were skipped). The aborted and skipped test cases would be highlighted with critical messages in the results.

It is a feature of Zonemaster that makes it possible to ignore or change the severity level. Your suggestion would require to change that.

I still think we should keep the test cases independent from each other, i.e. it should be possible to run any test case without first running another test case. It does not mean that it is always meaningful to run the test case, but the steps should be there.

>     1. Agree on the general outline of such an architecture before resolving this issue. (I laid out a rough idea above in this comment, based on suggestions put forth by you guys.)

Will you write an md file in PR?

>     2. Resolve this issue by merging an update to the specifications for how to deal with these broken DNSKEY records, that doesn't really do anything to implement the architectural changes, but that is roughly in line with the architecture we have in mind.

Where should that specification live and how should it be integrated with the rest.

>     3. Merge a refactoring of the specification for the broken DNSKEY behavior into a proper test case, along with any needed changes to the test case specification template, and along with any other changes needed to establish the architecture. Perhaps some new document; who knows, we'll see.

Is this the same as writing a new test case specification? Or what do you mean?

>     4. As part of our normal processes, incrementally bring existing test case specifications into line with the updated test case specification template.

Yes, we should update the test case specifications. Each time we do that, the test case implementations must also be updated, and we those resources.
